### ISSUE TITLE
`ParquetFile.from_pandas()` and more generally, implementation architecture?

### ISSUE THREAD

**Issue by yohplala** [dask/fastparquet#727]
Hello,
I would like to work out the ability to save data in Parquet format with fastparquet while working with vaex (that manages dataframes larger than RAM) in a 'convenient' way, meaning I will have vaex yielding pandas dataframes in chunk, only materializing in RAM pandas dataframes with specified chunk size.

For convenience, I would thus need to be able to submit to fastparquet writing function an iterable of pandas dataframes.  
this is something I have started to implement in PR #712 , with `ParquetFile.write_row_groups()` which accepts a dataframe or an iterable of dataframes indeferrently.
But the Parquet dataset initialization is missing.

I would like to propose thus to implement a `ParquetFile.from_pandas()` which would be a class method that is able
- to initialize a `Parquetfile` instance from a pandas dataframe or an iterable of pandas dataframes.
- in the case of an iterable of dataframe, we will need a `columns` parameter to provide the column index in parallel.
- similar to the other row group utilities that have been implemented so far, it would return the parquet file instance, and leave the ability to record or not common metadata (in case of `hive` file scheme).
It would mostly be the code already existing, that remains in `write()`.

This leads me to the topic of fastparquet architecture.
With this last function, fastparquet would then features:
- 'public' functions: `write`, `update`, `merge`. These high level functions rely for the most part on the 'yet-to-become-central' `ParquetFile` object.
- 'advanced' functions provided by `ParquetFile` class, which provide a more granular access to fastparquet inner functions.
- 'internal' functions with `ThriftObject`, encoding, utils, compression. Ideally, these functions would only be used within fastparquet lib.

The implementation of the row group utilities have lead me to see the organization of fastparquet in this way, and this would be strengthen with this new `ParquetFile.from_pandas()` class method.
Would the implementation of this last method to `ParquetFile` class be acceptable?

Thank you in advance for your feedback.
Bests,

**Comment by martindurant** [comment]
As ever, I would be happy to see your attempt at implementing this idea. *BUT*: `ParquetFile` is really, or at least was designed as, an interface for reading data. Yeas, indeed, we can attach methods to it to modify datasets, but it's more of a restructuring job to make it a full read-write thing, rather than the set of functions in the writer module.
What would the redesign gain for you?

The second major point, is that Dask already does what you are after. I don't know how vaex and dask might interact with one-another, but an obvious solution if you want parallel IO to parquet: use Dask. If you read he code in dask.dataframe.io.parquet, you will see that there is rather a lot of it, and that might give you some insight into how tricky it is to get everything right.

This does not mean I am opposed to creating an empty parquet dataset from metadata or piece of pandas, and then filling it by iterating over dataframe pieces. I'm not sure what a `from_pandas` would gain beyond a call to write followed by the current ParquetFile constructore.
Do bear in mind that parquet is designed to be parallel-friendly, and you are describing a serial approach.

On the broader architecture/restructuring question: it's quite a lot of work, especially when other libraries develop on fastparquet, so the motivation to carry it through would need to be very strong.

---

By the way, did you notice #721 ? That's a completely different possible future extension of fastparquet, one that doesn't use pandas at all. I haven't yet determined if it's worth the effort to do that, given that there is already a viable path using pyarrow, but with fewer features than we could provide with custom code here.

**Comment by yohplala** [comment]
> As ever, I would be happy to see your attempt at implementing this idea.

Thanks for this welcoming answer @martindurant. I will start implementing something, and if worthwhile, will submit it.

 > _BUT_: `ParquetFile` is really, or at least was designed as, an interface for reading data. Yeas, indeed, we can attach methods to it to modify datasets, but it's more of a restructuring job to make it a full read-write thing, rather than the set of functions in the writer module. What would the redesign gain for you?

The gain is probably little.
My thinking is that _fastparquet_ lib capitalizes efforts to make _Parquet_ format accessible in different ways from python. _BUT_ because `write` interface is already overcrowded with parameters, the proposal here is to split its job in 2 several smaller functions:
 - one, the append stuff, is the work done in #712 (`ParquetFile.write_row_groups()`)
 - the other, the parquet file creation from scratch, would be this new PR

These 2 sub-fonctions, by having a smaller workscope, would have a somewhat simpler interface, allowing for more features (possibly). In this context:
 - `ParquetFile.write_row_groups()` can sort row groups, and rename part files accordingly (as I understood you could suggest)
 - `ParquetFile.from_pandas()` could accept iterable of dataframes, with need to provide then `columns` as a separate parameter.

Attaching these sub-functions to `ParquetFile` class seems to me _logic_, with the ability to re-use `ParquetFile` instance. This may be little gain, hence me starting this paragraph with _"The gain is probably little."_

> The second major point, is that Dask already does what you are after. [...]  an obvious solution if you want parallel IO to parquet: use Dask.

There is a misunderstanding somewhere, possibly from you (about my intent), or possibly from me (about Dask).
 - regarding my intent
I am a person coding during my spare time, because I find the world of data thrilling and I have a personal project for which I need to manage "large size dataset" with respect to my hardware (I am sorry if that seems weird. I am absolutely not coding during my day job). Fitting to this context, my laptop has 4 cores (8 threads) and 16GB of RAM.
Nowadays, with _jit_ (like with _numba_) and correct python implementation you can process GB of data in few seconds. CPU is not my bottleneck. But RAM is. In 2017, Wes McKinney was writing _"Nowadays, my rule of thumb for pandas is that you should have 5 to 10 times as much RAM as the size of your dataset. So if you have a 10 GB dataset, you should really have about 64, preferably 128 GB of RAM if you want to avoid memory management problems."_ ([blog](https://wesmckinney.com/blog/apache-arrow-pandas-internals/))
**Takeway** I don't think I am CPU-bound, but memory-bound because of my hardware versus the dataset I am handling. I don't think to be looking for parallel IO to parquet.

> I don't know how vaex and dask might interact with one-another, [...]
> If you read he code in dask.dataframe.io.parquet, you will see that there is rather a lot of it, and that might give you some insight into how tricky it is to get everything right.

You are right that I should investigate what is done in Dask.
**Disclaimer** I am not the right person to compare Dask and Vaex. I have never used Dask (only read part of its documentation), and I am still very new to Vaex.
Both Dask and Vaex operate on _chunks_ and as such, both feature _graph management_, except not at the same scale
 - my understanding of Dask (and why I don't think it fits my need)
As per my understand, a chunk in Dask is a pandas dataframe ([A Dask DataFrame is [...] composed of many smaller Pandas DataFrames,](https://docs.dask.org/en/stable/dataframe.html)). When it operates in parallel, it loads several pandas dataframes in RAM to be processed by several cores. Default row group size in fastparquet being 50_000_000 rows and understanding (possibly wrong understanding here?) that a row group in fastparquet is a chunk in Dask, a 'typical' size order of a Dask chunk could be ~2,2GB (assuming arbitrarily 6 columns of int64/float64).
Given this size of chunk, and my hardware, I don't think I would really be able to scale my data processing over more than 2 cores by use of Dask. In this situation, pandas can do the job alone and avoiding Dask completely simplifies my implementation. (again, as per my understanding)
 - my understanding of Vaex (and why I think it is worth investigating versus my needs)
A chunk in Vaex varies between 1024 and 1024^2 rows. For 6 columns, this means a memory footprint between ~42kB & 48MB. The data source it operates can be loaded in RAM (a pandas dataframe or a numpy array for instance) or a file on disk allowing for memory mapping (arrow format or hdf5). In this context, I have enough RAM to make use of multiprocessing (even if this is not really what I am after as indicated above).

I was comparing both so as to introduce the beginning of an answer to your question  _"I don't know how vaex and dask might interact with one-another"_
**Takeaway** I think vaex is better compared to pandas (in terms of work scope), i.e. a library you can use to work on your own workstation. It seems to me Dask perimeter lies above, when it comes to manage graphs over several CPUs (either a 'big' workstation or a cluster).

Regarding my own use case, the logic is:
 - a row group in parquet is a pandas dataframe in memory (up to 6GB). I am iterating over these row groups. The IO work (parquet format on disk <-> memory) is managed with fastparquet.
 - for one given row group that is loaded in memory (one at a time) it can then either be processed
     - by use of pandas (less and less - take RAM, no multiprocessing)
     - or by use of vaex (more and more - does not take RAM, is multiprocessed)
     - or by use of Numba-boosted custom functions (when neither pandas nor vaex have the required vectorized functions available)

> This does not mean I am opposed to creating an empty parquet dataset from metadata or piece of pandas, and then filling it by iterating over dataframe pieces. I'm not sure what a `from_pandas` would gain beyond a call to write followed by the current ParquetFile constructor. Do bear in mind that parquet is designed to be parallel-friendly, and you are describing a serial approach.

Yes, you are right. I am still thinking that this function fits in my use case, but I do now see it does not necessarily does in Dask context. I will investigate, and submit only if it seems it is worthwhile. 

> On the broader architecture/restructuring question: it's quite a lot of work, especially when other libraries develop on fastparquet, so the motivation to carry it through would need to be very strong.

Thanks for the questions your answer raises. It helps me better understand what is at stake and why my 'understanding' is not necessarily appropriate. That is ok, I acknowledge this status.

> By the way, did you notice #721 ? That's a completely different possible future extension of fastparquet, one that doesn't use pandas at all. I haven't yet determined if it's worth the effort to do that, given that there is already a viable path using pyarrow, but with fewer features than we could provide with custom code here.

I could see the ticket, but I have not understood the implications. I will follow it and try better understanding.
You mentionned the `fastparquet.core.read_col` function, so I printed in my mind that this is for reading data (as opposed to writing). Is this related in some way to 'memory mapping'?

Thanks again for your answer Martin, much appreciated!

**Comment by martindurant** [comment]
A quick clarification for dask: it was designed first for a single-node situation such as a desktop machine or laptop. Typical chunk sizes for dataframe operations are ~100MB (in memory) - this is a recommendation that is still in the documentation. The large rows-per-row-group value for fastparquet represents a maximum, when the dataframe fits entirely in memory. Usually, Dask will write row-groups smaller than this, the size determined by wherever the data is coming from. Perhaps the default row_group_offsets should be smaller...

In short, yes Dask is designed for exactly the kind of workflow you are doing, and already handles the thorny bits of parquet IO. It operates in parallel *and* out-of-core, depending on your need, but also scales to very massive datasets, perhaps distributed on a cluster, and also workflows that are not dataframes too. The rule of thumb about RAM allocation is still valid: you should make available at least a few times the in-memory size of a partition times the number of partitions being processed at once. Yes, one partition in dask is loaded from at least one parquet row-group, although you can can select columns on load.

Caveat: I do not know much about vaex.

**Comment by yohplala** [comment]
> A quick clarification for dask: it was designed first for a single-node situation such as a desktop machine or laptop. Typical chunk sizes for dataframe operations are ~100MB (in memory) - this is a recommendation that is still in the documentation. The large rows-per-row-group value for fastparquet represents a maximum, when the dataframe fits entirely in memory. Usually, Dask will write row-groups smaller than this, the size determined by wherever the data is coming from. Perhaps the default row_group_offsets should be smaller...

Thanks for sharing your insight @martindurant .

As pointed before, I have not been really looking for parallelizing stuff.
When I started reading Dask documentation I could read in different places, this best practise (for instance in "[Best practise](https://docs.dask.org/en/latest/dataframe-best-practices.html)")
_"Use Pandas. For data that fits into RAM, Pandas can often be faster and easier to use than Dask DataFrame. While “Big Data” tools can be exciting, they are almost always worse than normal data tools while those remain appropriate."_

I may iterate over dataset sometime, but this is not systematic.
To be honest, at this point, vaex does not give reason to change, but maybe in the future I will give dask a 2nd try :)

More generally regarding this ticket, I am finally closing it.
I have been reviewing what could be done, and I do see that this refactoring work would not bring new significant features. I could also notice that dask is directly calling `make_metadata`, one of the main function in this part of the code. This makes any change complex to manage, and as just said, for no significant reason.

Thanks again @martindurant for taking the time to share your insight. Bests,
